### ISSUE TITLE
[Bug]: TENT deadline predictors disagree on bandwidth, and the drop predictor multiplies worker-queue wait by the request's slice count

### ISSUE THREAD

**Issue by SongOf** [kvcache-ai/Mooncake#3750]
### Bug Report

**Environment:** `main` (observed at 69c22878), TENT RDMA transport with the runtime queue enabled. No hardware needed to reproduce the arithmetic — it is in the code paths below.

**Summary**

A request carrying `deadline_ns` is judged by its *predicted MLU* (predicted completion time / remaining window). Two layers make a decision from that number, and each computes it from a different bandwidth through its own copy of the formula:

| Layer | Decision | Bandwidth used |
|---|---|---|
| `LocalTransferAdmissionQueue::pickForDispatch` → `shouldDrop` (RFC #2519 step 3, #2764/#2816) | drops the owner → `CANCELED` | `RdmaTransport::getEstimatedBandwidth()` = sum of every device's **selection EWMA** (`DeviceSelector::getAggregateEwmaBandwidth`) |
| `Workers::asyncPostSend` deadline arbitration (RFC #2792, #2794) | orders QP posting | the constant `transports/rdma/default_bandwidth_gbps` |

So the same request is ranked on two unrelated scales, and the layer that can cancel it stands on the weaker number. Three concrete defects:

**1. The selection EWMA folds worker-queue wait into "bandwidth", and the drop predictor then multiplies that wait by the request's slice count.**

`Workers::asyncPollCq` feeds `DeviceSelector::release()` with `overall_lat = poll_ts - enqueue_ts` per slice, i.e. queue wait `q` plus wire time. For a slice of length `l` on a link of rate `r` the learned "bandwidth" is `l / (q + l/r)`. `shouldDrop` predicts a request of `L` bytes as `L / bw`:

```
L / bw = L · (q + l/r) / l = (L/l) · q + L/r
```

The wait is charged once per slice, but a request's slices wait behind the same backlog concurrently — they wait once. A 64-slice request is predicted to wait 64×. Large requests are therefore systematically over-estimated and preferentially dropped, and the admission queue's own dispatch window (which is what fills the worker queue) feeds back into its drop decision.

**2. The series is ~99% the latest sample.** `bandwidth_learning_rate` defaults to 0.01 and weights the *old* value (`new = α·old + (1-α)·observed`), so a single slow completion can flip an irreversible drop.

**3. The arbitration's constant was never measured**, and after #3644 (per-NIC negotiated speed) it still reads the config default rather than the device's rate.

The selection EWMA cannot simply be "fixed": it includes queueing on purpose so that a backed-up NIC scores worse for device selection.

**What RFC #2519 actually specified**

> `available_bw` — derived from the existing `ewma_bandwidth_bps` **minus current in-flight load** on the pinned device.

The in-flight term was dropped when the predictor landed (#2764, #2816). The bug above is what happens when queueing is instead smuggled into the rate.

**Expected**

Both layers predict completion time the way `DeviceSelector` already scores devices — `(bytes_ahead + length) / wire_rate` — from one definition, with `bytes_ahead` as an additive term (what the request must wait behind: bytes already dispatched and not yet completed) over a queue-free wire rate, smoothed heavily enough that one outlier cannot cancel a user's transfer.

**Fix**

Prepared on `SongOf:fix/tent-mlu-bandwidth-estimate`, stacked on #3644 (it needs the per-device link speed as the seed and the availability flag for the aggregate):

- one predictor, `DeadlineMlu(bytes_ahead, length, deadline_ns, now_ns, bw_bps)` in `tent/runtime/deadline_mlu.h`, called by both `shouldDrop` and `PredictedMlu`;
- a per-device **transmit estimate** (`ewma_transmit_bps`): same update rule, clamp and link-speed seed as the selection EWMA, but fed post→completion time only and smoothed with its own `transmit_bandwidth_learning_rate` (default 0.99); the selection EWMA is unchanged;
- `bytes_ahead` = the admission queue's `dispatching_bytes_` (+ what the current pick already took) for the drop, and the NIC's inflight bytes for the arbitration;
- `getEstimatedBandwidth()` returns the aggregate transmit estimate; the arbitration reads the local NIC's.

Tests reproduce each point without hardware (same samples → selection EWMA 1.55 GB/s vs transmit 3.1 GB/s; a 30 GB/s outlier moves the selection EWMA above 25 GB/s but the transmit estimate by ~1%; a 16 B / 500 µs-window owner is dispatched on an idle queue, dropped behind a 1 MB owner in flight, and dispatched again once it completes). PR to follow once #3644 merges.

Refs #2519 (restores the in-flight term), #2764, #2816, #2794, #2132.

---

### Follow-up left open by #3644 — addressed by #3777

The predictor above divides by a per-NIC bandwidth that #3644 seeds from the negotiated port speed and re-reads on port events. Its review left one input out of scope:

**LAG-effective speed (`ibv_query_port_speed()`).** `linkSpeedGbps()` decoded `active_speed × active_width`, the physical link's encoded rate. For a VF over VF LAG / multiport LAG, a PF dropping out of the bond halves the VF's effective bandwidth while its port stays `ACTIVE` at the same encoding, so the re-read on `IBV_EVENT_DEVICE_SPEED_CHANGE` saw `after == before` and did not re-seed. rdma-core v62 added `ibv_query_port_speed()` alongside the event to report the *effective* speed (100 Mb/s units).

**#3777** resolves it as an optional `IbvLoader` symbol (older libraries keep RDMA and the encoded rate), prefers it in `linkSpeedGbps()` via `ibPortSpeedGbps()`, and adds a verbs-injection seam on `RdmaContext` so the effective-speed and refresh paths are tested without an RNIC.

### Before submitting...

- [x] Ensure you searched for relevant issues and read the documentation

**Comment by he-yufeng** [comment]
Checked this against current main after #3511 landed (2026-09-02), because it touches exactly the sampling side of defect 1.

What is already fixed: the EWMA sample in `asyncPollCq` is now the current attempt's inflight time (`poll_ts - submit_ts`), gated on `IBV_WC_SUCCESS`, instead of cumulative since-enqueue latency. So worker-queue wait and previous failed attempts no longer enter the learned bandwidth, and the "queue wait multiplied by slice count" effect you describe does not exist anymore in that form.

What still stands: the two layers still judge the same request on unrelated scales.

- `LocalTransferAdmissionQueue::pickForDispatch`/`shouldDrop` reads `bandwidth_provider_()`, which resolves to `RdmaTransport::getEstimatedBandwidth()` = `DeviceSelector::getAggregateEwmaBandwidth()`.
- `Workers::asyncPostSend` deadline arbitration sorts by a predicted MLU computed from the constant `transports/rdma/default_bandwidth_gbps`.

So a drop decision can fire on the learned number while the posting order is computed on the configured constant, and the two can diverge arbitrarily (EWMA collapsed by a slow path vs a static default). The dropping layer also estimates whole-request time as `length / bw_bps` with no model of the slice schedule, so per-request parallelism is invisible to it either way.

My suggestion: make the arbitration read the same learned scale the drop predictor uses (or give both layers the EWMA), and document which scale is authoritative for deadline math. Unifying on the constant instead would also work but throws away the learning #3511 just made honest. I can send the change if a maintainer confirms the direction; it is a small diff either way, the design choice is the real question.

**Comment by SongOf** [comment]
@he-yufeng https://github.com/kvcache-ai/Mooncake/pull/3780 I've already implemented it in that PR; if you're interested, feel free to offer suggestions.
